Rails の i18n を理解するの課題提出

yokomaru opened the pull request:
# 必須要件
- [x] 英語と日本語の翻訳をconfig/locales配下のYAMLファイルに定義する
- [x] 各画面の表示を翻訳ファイルから取得して表示できるようにする
- [x] 表示を切り替えるときは「アプリケーションの設定ファイルを変更してサーバー再起動」でOK
    - [x] 提出時は日本語をデフォルトにすること）
- [x] Rails 国際化 (i18n) API - Railsガイド を読み、使えそうなメソッドがあれば積極的に使う
- [x] 町田スペシャルCSSを適用してUIを見やすくする

# スクリーンショット
- [x] デフォルトの表示言語を英語または日本語に切り替えることで、実際に画面の表示が英語と日本語に切り替わる
- [x] rubocopとerblintをパスさせる
## 以下提出物にて貼り付け済み
- https://bootcamp.fjord.jp/products/20048/

# 歓迎要件
- [x] 画面上のリンクやURLパラメータで英語と日本語を切り替えられる

yokomaru committed:
first commit

yokomaru committed:
add: scafolds.css

machidaスペシャルCSSファイルを追加

yokomaru committed:
update: erb files for applying css

machidaスペシャルCSSを以下erbファイルに適用
・edit.html.erb
・index.html.erb
・new.html.erb
・show.html.erb

yokomaru committed:
update: .gitignore

.vscode/launch.jsonの設定を追加したので .gitignoreに追加

yokomaru committed:
add: gem i18n_generators

yokomaru committed:
add: write config.i18n informations to application.rb

application.rbにi18nの情報を追加

yokomaru committed:
add: ja.yml

- [svenfuchs/rails\-i18n: Repository for collecting Locale data for Ruby on Rails I18n as well as other interesting, Rails related I18n stuff](https://github.com/svenfuchs/rails-i18n) を参考にja.ymlを追加
- model book を追加
- linkの表示文字の翻訳文を追加

yokomaru committed:
update: link and title

翻訳対応のため各linkとtitleの文字をtメソッドで囲む
 Book.model_name.humanでtitleを表示

yokomaru committed:
update: add Book.human_attribute_name

書籍の表示時の各項目に翻訳対応

yokomaru committed:
add: I18n.available_locales to apprication.rb

I18n使用可能なロケールをjaとenに固定

yokomaru committed:
update: separate  locale file directoties

ロケールファイルを以下ディレクトリに分ける
- App共通
   - config/locales
- Model
   - config/locales/models
- View
   - config/locales/views

yokomaru committed:
update: apply sepalate locale files

- erbファイルに分割したロケールファイルを読み込むように修正

yokomaru committed:
fix: locale file

- 不要な項目の削除
- 最後の改行の追加

yokomaru committed:
add: model message to model locale files

- 共通の作成・更新・削除の成功メッセージをmodelのロケールファイルに追加
- contollerに適用

yokomaru committed:
update: change yaml hierarchy of view locale file

ビューのロケールファイルの階層を修正
上記に伴いビューファイルから呼び出す箇所も修正

yokomaru committed:
fix: locale en

yokomaru committed:
add: title new to view locale file

- 追加漏れのためビューファイル用のロケールファイルに新規画面を追加
- 新規作成画面にr10nを適用

yokomaru committed:
add: change language button and param

- 画面上に言語切り替えボタンを追加(urlからも変更可能)
- ApplicationControllerにlocaleのparamをセットする処理を追加

yokomaru committed:
delete: vscode launch.json .gitignore

yokomaru committed:
fix: yml first row

他のymlファイルと統一して行等の---を削除

yuuu commented on app/views/books/new.html.erb:
```diff
@@ -1,9 +1,9 @@
-<h1>New book</h1>
+<h1><%= t('common.title.new', model: Book.model_name.human) %></h1>
```
こういった特定のページでのみ使用する翻訳は [遅延探索](https://railsguides.jp/i18n.html#%E9%81%85%E5%BB%B6%E6%8E%A2%E7%B4%A2) を使うようにしましょう。記載もスッキリしますし、どこに翻訳が書かれているかエンジニア間で認識共有しやすくなります 👍

yokomaru replied:
こちら以下修正しました！
戻るボタンと詳細への遷移ボタンは複数箇所使用しているところがあったので、
共通用のロケールファイルに追加して、`views.common.back`のように呼び出しています。

[f42ebad](https://github.com/yokomaru/fjord-books_app-2023/pull/1/commits/f42ebad16f1791a6216867ea4bbe6b22db7644da)
[fd70335](https://github.com/yokomaru/fjord-books_app-2023/pull/1/commits/fd70335f42ad6911bcf55ac608410dd21cfb6f13)

yuuu commented on app/views/layouts/application.html.erb:
```diff
@@ -9,8 +9,15 @@
   </head>
-
+  <div class="nav" style="text-align: right;">
+    <% if I18n.locale.to_s == "en" %>
```
ここは2つの観点で修正いただきたいです。

- 今の実装だと英語を読む人に向けに「日本語」と表示していることになります。「Japanese」と表示した方が親切ですね。
    - Japaneseや英語もi18n対応するようにしましょう。そうしておけばこの問題にも気づきやすくなります。
- Viewにロジックが入ってしまっているので、if式ごとHelperに移動させることを検討したいです

yokomaru replied:
> 今の実装だと英語を読む人に向けに「日本語」と表示していることになります

た、確かに・・！気づいておりませんでしたっ🙏
ありがとうございます！

こちらを実装するにあたり、実装の仕方で悩んでいるので相談させてください！

## 実現したいこと
- localeがjaの時は切り替えボタンを英語で表示したい(Englishと表記)
- localeがenの時は切り替えボタンを日本語で表示したい(日本語と表記)

- イメージ
<img width="880" alt="スクリーンショット 2024-06-21 22 09 01" src="https://github.com/yokomaru/fjord-books_app-2023/assets/44030266/dd7a2bfb-812f-437c-aa14-95851b6e9b6a">
 

## 対応していること
- [x] books_helper.rbに処理を移す
- [x] 共通用のlocaleファイルにjapanese、englishというキーを追加

## 現状の実装
- app/helpers/books_helper.rb
```ruby
# frozen_string_literal: true

module BooksHelper
  def switch_language
    if I18n.locale.to_s == "en"
      link_to(t('views.common.japanese'), locale: 'ja')
    else
      link_to(t('views.common.english'), locale: 'en')
    end
  end
end

```

- app/config/locales/views/en.yml
```yml
en:
  views:
    common:
      show: "Show This %{model}"
      back: "Back to %{model}s"
      japanese: Japanese
      english: English
```

- app/config/locales/views/ja.yml
```yml
ja:
  views:
    common:
      show: "この%{model}の詳細を見る"
      back: "%{model}の一覧へ戻る"
      japanese: 日本語
      english: 英語
```

## 相談したいこと
- localeがjaの時に、en.ymlのファイルを読み込みたい(逆も然り)
- helperの分岐で以下のように`I18n.t()`メソッドを使っているが、現状のロケールのファイルを読み込みに行ってしまうので実現できていない。

```ruby
# 日本語 と表記してほしいがJapaneseとなってしまう
 link_to(t('views.common.japanese'), locale: 'ja')
```

- 以下のようにトップのキー渡すと、`ja.en.views.common.japanese`を探してしまうのでmissing templateエラーになってしまう。

```ruby
 link_to(t('en.views.common.japanese'), locale: 'ja')
```

- localeを使用した状態で上記を実現できる方法があれば教えていただきたいです🙇‍♀️
    - [Rails 国際化（I18n）API \- Railsガイド](https://railsguides.jp/i18n.html)を読んで使えそうなメソッドなどを探してみたのですが見つからず。。！

## 考えた対応
### ja.ymlとen.ymlの表示を以下のようにする
- 実現はできるけど名前と実態が伴っていないので微妙だと感じている

- app/config/locales/views/en.yml
```yml
en:
  views:
    common:
      show: "Show This %{model}"
      back: "Back to %{model}s"
      japanese: 日本語
      english: English
```

- app/config/locales/views/ja.yml
```yml
ja:
  views:
    common:
      show: "この%{model}の詳細を見る"
      back: "%{model}の一覧へ戻る"
      japanese: 日本語
      english: English
```

### helperに値を直書きする
- 一番楽に実現できるがメンテナンス性に欠ける

```ruby

module BooksHelper
  def switch_language
    if I18n.locale.to_s == "en"
      link_to('日本語', locale: 'ja')
    else
      link_to('English', locale: 'en')
    end
  end
end
```

yokomaru replied:
すみません、上記改めて見直したところ私の認識が間違っていました・・！

## コメントいただいた内容
- localeがjaの時は切り替えボタンを日本語で表示(英語と表記)
- localeがenの時は切り替えボタンを日本語で表示(Japaneseと表記)
→ 英語を読む人には英語表記、日本語で読む人には日本語で表記する

## 自分が実装しようとしていたこと
- localeがjaの時は切り替えボタンを英語で表示(Englishと表記)
- localeがenの時は切り替えボタンを日本語で表示(日本語と表記)
→ 英語を読む人がlocale=jaのページに飛んでしまった時に、ロケールが変えやすいように切り替えボタンをEnglishと表示させる必要があると思った

上記を改めて見比べたとき、ご指摘いただいた実装の方がシンプルでわかりやすいと感じました！
実装はできているので、こちらで改めて提出いたしまします。
お騒がせしました！

yuuu commented on config/locales/en.yml:
```diff
@@ -1,33 +1,216 @@
-
 en:
-  hello: "Hello world"
+  activerecord:
```
`rails-i18n` というgemを導入するとこのあたりの翻訳は記載の必要がなくなります 👍

yokomaru replied:
ありがとうございます！
こちら`rails-i18n`を追加しました！

yokomaru committed:
add: gem 'rails-i18n' and delete default ja.yml

gem 'rails-i18n'を追加し、不要になったymlファイルを削除

yokomaru committed:
update: book.yml

- booksに関するページは遅延探索ができるようにディレクトリを分けた上で各ビューごとにローカライズしたいキーを記載し、遅延探索を利用できるようにする
- 共通で使っている戻るボタンは/locale/ymlにcommonとして記載しviews.common.backで呼び出す

yokomaru committed:
fix: chenge locale key show books to common

複数箇所使用している詳細ボタンへのリンクをcommon用のlocaleファイルに移す

yokomaru committed:
add: switch_language method

- htmlに書いていた分岐をbook_helper.rb移動
- yamlに翻訳を追加

yokomaru committed:
fix: rubocop
